internal/pkg/db/db.go

```diff
@@ -93,14 +94,79 @@ func (backend *Backend) AddDocument(document *sbom.Document) error {
 func (backend *Backend) GetDocumentByID(id string) (*sbom.Document, error) {
 	document, err := backend.Retrieve(id, nil)
 	if err != nil {
-		backend.Logger.Warn("Document could not be retrieved", "id", id, "err", err)
+		backend.Logger.Debug("Document could not be retrieved", "id", id, "err", err)
 
 		return nil, fmt.Errorf("failed to retrieve document: %w", err)
 	}
 
 	return document, nil
 }
 
+func (backend *Backend) GetDocumentByIDOrAlias(id string) (*sbom.Document, error) {
+	document, idErr := backend.GetDocumentByID(id)
+
+	if document == nil {
+		documents, aliasErr := backend.GetDocumentsByAnnotation("alias", id)
+		if aliasErr != nil {
+			backend.Logger.Debug("Document could not be retrieved", "alias", id, "err", aliasErr)
+
+			return nil, fmt.Errorf("failed to get document by ID or alias: %w | %w", idErr, aliasErr)
+		}
+
+		document = documents[0]
+	}
+
+	return document, nil
+}
+
+func (backend *Backend) GetDocumentsByIDOrAlias(ids ...string) ([]*sbom.Document, error) {
+	if len(ids) == 0 {
+		documents, err := backend.GetDocumentsByID()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get documents by ID: %w", err)
+		}
+
+		return documents, nil
+	}
+
+	documents := make([]*sbom.Document, len(ids))
+
+	for i, id := range ids {
+		document, err := backend.GetDocumentByIDOrAlias(id)
+		if err != nil {
+			return nil, err
+		}
+
+		documents[i] = document
+	}
+
+	return documents, nil
+}
+
+func (backend *Backend) FilterDocumentsByTag(documents []*sbom.Document, tags ...string) ([]*sbom.Document, error) {
+	taggedDocuments, err := backend.GetDocumentsByAnnotation("tag", tags...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get documents by tag: %w", err)
+	}
+
+	taggedDocumentIDs := make([]string, len(taggedDocuments))
+	for i, taggedDoc := range taggedDocuments {
+		taggedDocumentIDs[i] = taggedDoc.Metadata.Id
+	}
```

Comment:
If looping and assigning this way, prefer to see

- rename `i` ---> `idx` to align with project convention
- just use `for idx := range taggedDocuments` and reference as `taggedDocuments[idx].Value`

Reply:
Also can you describe the optimization this implements, specifically?

Reply:
The (small) optimization is just preallocating the space in the slice and then assigning the values on each index instead of dynamically appending to the slice.